meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+collectNeededInputsFromObservation (ValueGE value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (ValueGT value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (ValueLT value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (ValueLE value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (ValueEQ value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (TrueObs) = S.empty
+collectNeededInputsFromObservation (FalseObs) = S.empty
+
+collectNeededInputsFromContract :: Contract -> S.Set IdInput
+collectNeededInputsFromContract Null = S.empty
+collectNeededInputsFromContract (Commit _ _ _ value _ _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromValue value
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Pay _ _ _ value _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromValue value
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Both contract1 contract2) =
+  S.unions [ collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Choice observation contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (When observation _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (While observation _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Scale value1 value2 value3 contract) = 
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2
+           , collectNeededInputsFromValue value3
+           , collectNeededInputsFromContract contract ]
+collectNeededInputsFromContract (Let _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Use _) = S.empty
```

Comment:
I'll just leave these pamphlets here and you can look at them when you're ready.

Reply:
:+1: